File: Homomorphic Encryption – Process Encrypted Data Without Decryption.go
package main

import (
	"fmt"
	"math/big"

	"github.com/tuneinsight/lattigo/v4/bfv"
	"github.com/tuneinsight/lattigo/v4/rlwe"
)

func main() {
	// Step 1: Create encryption parameters
	params, err := bfv.NewParametersFromLiteral(bfv.PN12QP109)
	if err != nil {
		panic(err)
	}

	// Step 2: Generate keys
	kgen := bfv.NewKeyGenerator(params)
	sk, pk := kgen.GenKeyPair()
	encryptor := bfv.NewEncryptor(params, pk)
	decryptor := bfv.NewDecryptor(params, sk)
	evaluator := bfv.NewEvaluator(params, rlwe.EvaluationKey{})

	// Step 3: Encrypt two integers
	encoder := bfv.NewEncoder(params)
	plaintext1 := bfv.NewPlaintext(params)
	plaintext2 := bfv.NewPlaintext(params)

	encoder.Encode([]uint64{10}, plaintext1)
	encoder.Encode([]uint64{20}, plaintext2)

	ciphertext1 := encryptor.EncryptNew(plaintext1)
	ciphertext2 := encryptor.EncryptNew(plaintext2)

	fmt.Println("🔒 Numbers Encrypted Successfully!")

	// Step 4: Perform Homomorphic Addition (E(10) + E(20) = E(30))
	ciphertextSum := evaluator.AddNew(ciphertext1, ciphertext2)

	// Step 5: Decrypt the result
	decryptedResult := bfv.NewPlaintext(params)
	decryptor.Decrypt(ciphertextSum, decryptedResult)

	var result []uint64
	encoder.Decode(decryptedResult, &result)

	fmt.Println("🔓 Decrypted Sum:", result[0]) // Should print 30
}

File: Homomorphic Encryption – Process Encrypted Data Without Decryption.rs
use concrete::*;

fn main() -> Result<(), CryptoAPIError> {
    // Step 1: Create Encryption Parameters
    let config = ConfigBuilder::new()
        .enable_default_parameters()
        .build();

    // Step 2: Generate Keys
    let keygen = KeyGenerator::new(config);
    let (client_key, server_key) = keygen.generate();

    // Step 3: Encrypt two numbers
    let encrypted_num1 = client_key.encrypt(10)?; // Encrypt 10
    let encrypted_num2 = client_key.encrypt(20)?; // Encrypt 20

    println!("🔒 Numbers Encrypted Successfully!");

    // Step 4: Perform Homomorphic Addition (E(10) + E(20) = E(30))
    let encrypted_sum = server_key.add(&encrypted_num1, &encrypted_num2)?;

    // Step 5: Decrypt the result
    let decrypted_result = client_key.decrypt(&encrypted_sum)?;

    println!("🔓 Decrypted Sum: {}", decrypted_result); // Should print 30

    Ok(())
}
